chore(codecs): Add RFC for "Framing and Codecs - Sinks" (#9864)
vectordotdev/vector: rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

# RFC 8621 - 2021-10-29 - Framing and Codecs - Sinks

This RFC discusses changes to apply framing and encoding across sinks in a consistent way. On a high-level, this feature aims to be the symmetric counter part to the concept outlined [Framing and Codecs - Sources]([/blob/master/rfcs/2021-08-06-8619-framing-and-codecs-sources.md](https://github.com/vectordotdev/vector/blob/7796b3e766085225d2ebbe698a43d4015fe303c5/rfcs/2021-08-06-8619-framing-and-codecs-sources.md)).

## Context

> In the context of sinks, we refer with _encoding_ to serializing an event to bytes and with _framing_ to the process of wrapping one or multiple serialized events into a bounded message that can be sent as payload to a sink.

Author: 💯 for suggestions of a native speaker 😄

Furthermore, the shared `EncodingConfig` as it is implemented today is concerned with three tasks: Reshaping an event (including/excluding fields), serializing an event to a byte message, and framing/batching events. In accordance with the functionality that the decoding side provides, we want keep these concepts separate for a simpler mental model. In fact, we want to separate reshaping from codecs entirely and move it to the responsibilities of the schema work as outline in the next section.

## Cross cutting concerns

The codec work is very interweaved with the ongoing [schema work](https://github.com/vectordotdev/vector/pull/9388). It introduces the concept of "mappers" which understand how to transform an event at runtime based on the schema from its source and destination within Vector's topology.

We want to hand over the reshaping responsibilities to the schema, and provide a transition layer to support existing reshaping capabilities of sinks until schema mappers are implemented.

> Another related current work is `StandardEncoding` which has been introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent set of codec implementations aims and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.

Reviewer: ```suggestion
Another related change is `StandardEncodings` which was introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent implementation of common codecs, and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.

- Strategy for integrating encoding/framing with sinks

### Out of scope

- Integration with schema system (automatic transformation of event fields depending)

## Pain

Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.

Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.

## Proposal

### User Experience

Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.

> The fields containing event transformations in sinks on the current `encoding` options (`schema`, `only_fields`, `except_fields`, `timestamp_format`) should be moved to a dedicated option, e.g. `schema` or `transform`. Thus, this would introduce a breaking change in configurations. However, migration would be relatively straight-forward by nesting these options under the new key.

Reviewer: The one question that I think needs to be answered here is: will the schema work land before this happens?  If it doesn't, then we probably need a way handle parsing these out of the existing `encoding` configuration field.  We could deprecate setting them via `encoding` at the same time, but we would have to provide a deprecation window regardless.

Reviewer 2: Yeah, I would suggest we continue to support the current config interface for now and focus on organizing the internal implementation for now. Then once both this work and the schema work is further along we can tackle changing the external API with a bit more confidence.

Author: Sounds like a good plan 👍 Should be able to support existing config options without carrying over too much baggage.

Author: I came up with a wrapper that is able to apply the existing transformation behavior to the new encoding options while being fully backwards compatible to the old sink-specific encoding options: https://github.com/vectordotdev/vector/pull/10253.

For usage, have a look at https://github.com/vectordotdev/vector/pull/10684.

Author: Clarified in 21f9c36d82fdc5cbe8a299f7c7677a35bfd4ae24.

### Implementation

The following common config structures will be provided, analogously to the decoding/framing configs:

```rust
pub trait SerializerConfig: Debug + DynClone + Send + Sync {
    /// Builds a serializer from this configuration.
    ///
    /// Fails if the configuration is invalid.
    fn build(&self) -> crate::Result<BoxedSerializer>;
}
```

```rust
pub trait FramingConfig: Debug + DynClone + Send + Sync {
    /// Builds a framer from this configuration.
    ///
    /// Fails if the configuration is invalid.
    fn build(&self) -> crate::Result<BoxedFramer>;
}
```

> These can be build to form an `Encoder`:

Reviewer 2: ```suggestion
These can be built and combined to form an `Encoder`:
```

```rust
/// An encoder that can encode structured events to byte messages.
pub struct Encoder {
    serializer: BoxedSerializer,
    framer: BoxedFramer,
}

> ```

Reviewer: One thing I'd like to see here is if we can actually make `SerializerConfig`/`FramingConfig` provide their outputs generically.  For example:

```rust
pub trait SerializerConfig {
    type Serializer;
    
    fn build(&self) -> crate::Result<Self::Serializer>;
}

pub trait FramingConfig {
    type Framer;
    
    fn build(&self) -> crate::Result<Self::Framer>;
}

pub struct Encoder<S, F> {
    serializer: S,
    framer: F,
}
```

This is sort of an implementation detail, but we should avoid boxing individual components where possible, and prefer to box at a higher level.  It might even be possible to entirely avoid boxing.  I'm bringing this up here on the off-chance that not using boxed inner fields would meaningfully change how we have to approach the implementation.

Since codecs touch every single event entering and exiting Vector, we should take care to avoid boxing where possible.

Author: Hmm, this is a very interesting point.

Since the source side is structured very similarly, I tried to apply your suggestion to the `FramingConfig` and `ParserConfig`.

It fails to compile since `dyn_clone::clone_trait_object` needs to specify the associated type. Cloning a decoder/encoder is a rather common operation, it happens for each message or stream to reset the codec state.

I'm not sure if it's possible to circumvent this without having a `Box` on the inner type, but I'd be happy for any suggestions.

> `Encoder` implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<SmallVec<[Event; 1]>>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.

Reviewer 2: Why have `Encoder` work on `SmallVec<Event>` instead of just `Event`? I don't think we have the came concern as decoding where we can pull out multiple events.

Author: Dropped it in b2730814d5537ccd0710cc438b99e24fd309efbc. I think I came up with that at first since I assumed it would affect how the framing boundaries are serialized. However, synchronously waiting to collect events would be pretty bad. We'll need to constrain the framers on the encoding side to always produce the next serialized chunk when an event is encoded.

## Plan Of Attack

Incremental steps to execute this change. These will be converted to issues after the RFC is approved:

> - [ ] Implementation of support structures (`Encoder`, `SerializerConfig`, `FramingConfig`)

Reviewer 2: Probably worth mentioning some kind of backwards-compat layer here so that we can transition the actual user-facing config API on a separate cadence from the implementation.

Author: Added a note in 7e087fc5a526bc13d1511a959af7263efd866d46

> - [ ] Implementation of selected `encoders`/`framers` (e.g. `JSON` and `newline_delimited`)

Reviewer: Are you envisioning that we'll take the existing parsers and update them to also be able to encode, or will "parsers" and "encoders" always be separate?

Author: We had a related discussion about it when implementing the source side: https://github.com/vectordotdev/vector/pull/7482#discussion_r647880464.

I think the behavior for both source/sink framers and parsers/encoders should be symmetric enough to combine them.

Author: At this point, I'm not convinced any more that it's clever to reuse the same structs to implement deserializing and serializing, since they sometimes have internal state or configuration options that are only relevant for either side.

However, colocating the implementations for both sides seems worthwhile.

Opened #10087 to adjust the naming scheme, such that it won't clash when we implement the encoding side.

|`azure_blob`| `EncodingConfig<Encoding { Ndjson, Text }>` | Enveloped in EncodedEvent<PartitionInnerBuffer>. `Text` reads `message_key()`
|`azure_monitor_logs`| `EncodingConfigWithDefault<Encoding { Default }>` | Reshapes events only, without encoding
|`blackhole`| -/-
|`clickhouse`| `EncodingConfigWithDefault<Encoding { Default }>` | Reshapes events only, without encoding
|`console`| `EncodingConfig<Encoding { Text, Json }>` | `Text` reads `message_key()`
|`datadog`| `EncodingConfigFixed<DatadogLogsJsonEncoding>` | Doesn't provide options to encode the event payload separately from the protocol
|`datadog_archives`|
|`elasticsearch`|
|`file`| `EncodingConfig<Encoding { Text, Ndjson }>` | `Text` reads `message_key()`
|`gcp`|
|`honeycomb`|
|`http`| `EncodingConfig<Encoding { Text, Ndjson, Json }>` | Enveloped in HTTP request. Request-level compression. Sets headers depending on encoding config
|`humio`|
|`influxdb`|
|`kafka`| `EncodingConfig<StandardEncodings { Text, Json, Ndjson }>` | Doesn't reshape, uses `Encoder<Event>` for `StandardEncodings` in `encode_input`, enveloped in `KafkaRequest`
|`logdna`|

> |`loki`|

Reviewer 3: The `loki` sink supports encoding, doesn't it?

https://github.com/vectordotdev/vector/blob/d3a4492ee7f250c30c6e26592a7d8d4fdb3de70f/src/sinks/loki/config.rs#L18

> |`nats`|

Reviewer 3: The `nats` sink supports encoding, doesn't it?

https://github.com/vectordotdev/vector/blob/d3a4492ee7f250c30c6e26592a7d8d4fdb3de70f/src/sinks/nats.rs#L194-L205

|`new_relic_logs`|
|`papertrail`|
|`pulsar`|
|`redis`|
|`sematext`|
|`socket`| `EncodingConfig<Encoding { Text, Json }>` | `Text` reads `message_key()`

> |`splunk_hec`|

Author: Filled the blanks with my recent changes

# RFC 8621 - 2021-10-29 - Framing and Codecs - Sinks

This RFC discusses changes to apply framing and encoding across sinks in a consistent way. On a high-level, this feature aims to be the symmetric counter part to the concept outlined [Framing and Codecs - Sources]([/blob/master/rfcs/2021-08-06-8619-framing-and-codecs-sources.md](https://github.com/vectordotdev/vector/blob/7796b3e766085225d2ebbe698a43d4015fe303c5/rfcs/2021-08-06-8619-framing-and-codecs-sources.md)).

## Context

In the context of sinks, we refer to _encoding_ as serializing an event to bytes, and to _framing_ as the process of wrapping one or more serialized events into a bounded message that can be used as the payload for a request sent by a sink.

Currently, most sinks include the common `EncodingConfig<T>` in their config. It takes a generic argument where enums specify which encodings they support. However, the actual encoding logic is reimplemented by each sink individually, rather than falling back to a shared inventory of codec implementations. This leads to unnecessary drift in their feature set and behavior.

> Furthermore, the shared `EncodingConfig` as it is implemented today is concerned with three tasks: Reshaping an event (including/excluding fields), serializing an event to a byte message, and framing/batching events. In accordance with the functionality that the decoding side provides, we want keep these concepts separate for a simpler mental model. In fact, we want to separate reshaping from codecs entirely and move it to the responsibilities of the schema work as outline in the next section.

Reviewer 4: ```suggestion
Furthermore, the shared `EncodingConfig` as it is implemented today is concerned with three tasks: Reshaping an event (including/excluding fields), serializing an event to a byte message, and framing/batching events. In accordance with the functionality that the decoding side provides, we want to keep these concepts separate for a simpler mental model. In fact, we want to separate reshaping from codecs entirely and move it to the responsibilities of the schema work as outlined in the next section.
```

Another related change is `StandardEncodings` which was introduced in [#9215](https://github.com/vectordotdev/vector/pull/9215). It provides a consistent implementation of common codecs, and therefore aligns with our goals. However, it does not separate the reshaping responsibilities.

## Scope

### In scope

- Common support structures for encoding/framing
- Set of commonly used encoding (text, JSON, ...) and framing (newline-delimited, octet-counting, ...) methods
- Strategy for integrating encoding/framing with sinks

### Out of scope

- Integration with schema system (automatic transformation of event fields depending)

## Pain

Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.

Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.

## Proposal

### User Experience

> Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key.

Reviewer 2: Should they be able to set these on every type of sink? I'm curious if it would ever make sense to set `framing` on something like the `datadog_logs` sink.

Author: Clarified in 6c19b6bf7b80b9adc1971cae75db08d5f4f90b83.

|`aws_kinesis_firehose`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `rusoto_firehose::Record` that serializes to base64. `Text` reads `message_key()` | -
|`aws_kinesis_streams`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `rusoto_kinesis::PutRecordsRequestEntry`. `Text` reads `message_key()`
|`aws_s3`| `EncodingConfig<StandardEncodings { Text, Json, Ndjson }>` | ✔︎ | Uses `util::{RequestBuilder, Encoder, Compressor}`. `Text` reads `message_key()`
|`aws_sqs`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | Enveloped in `EncodedEvent<SendMessageEntry>`. `Text` reads `message_key()`
|`azure_blob`| `EncodingConfig<Encoding { Ndjson, Text }>` | ✔︎ | Enveloped in `EncodedEvent<PartitionInnerBuffer>`. `Text` reads `message_key()`
|`azure_monitor_logs`| `EncodingConfigWithDefault<Encoding { Default }>` | ✔︎ | Serializes to JSON. Enveloped in HTTP request
|`blackhole`| - | - | -
|`clickhouse`| `EncodingConfigWithDefault<Encoding { Default }>` | ✔︎ | Serializes to JSON. Enveloped in HTTP request
|`console`| `EncodingConfig<Encoding { Text, Json }>` | ✔︎ | `Text` reads `message_key()`

> |`datadog`| `EncodingConfigFixed<DatadogLogsJsonEncoding>` | ✔︎ | Doesn't provide options to encode the event payload separately from the protocol

Reviewer 5: There is no `datadog` sink. This is describing the `datadog_logs` sink. The `datadog_events` sink is missing, which is one of the more interesting ones. It doesn't allow setting the encoding at all, and uses the `EncodingConfig` to filter out fields that the datadog events api doesn't accept.

The fields containing event transformations in sinks on the current `encoding` options (`only_fields`, `except_fields`, `timestamp_format`) should also work with the new encoding options without breaking backwards compatibility. In the future, they may be moved to a dedicated key, e.g. `transform` or replaced by a mechanism provided by the upcoming schema support.

### Implementation

The following common config structures will be provided, analogously to the decoding/framing configs:

```rust
pub trait SerializerConfig: Debug + DynClone + Send + Sync {
    /// Builds a serializer from this configuration.
    ///
    /// Fails if the configuration is invalid.
    fn build(&self) -> crate::Result<BoxedSerializer>;
}
```

```rust
pub trait FramingConfig: Debug + DynClone + Send + Sync {
    /// Builds a framer from this configuration.
    ///
    /// Fails if the configuration is invalid.
    fn build(&self) -> crate::Result<BoxedFramer>;
}
```

These can be built and combined to form an `Encoder`:

```rust
/// An encoder that can encode structured events to byte messages.
pub struct Encoder {
    serializer: BoxedSerializer,
    framer: BoxedFramer,
}
```

> `Encoder` implements `tokio_util::codec::Encoder<Event>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<Event>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.

Reviewer 2: Do you imagine this will work in the same way for sinks that don't use one of the pieces (e.g. framing)?

It seems like sinks that entirely do their own serialization and framing (e.g. datadog_logs) would have an easy enough time opting out entirely, but would something like kafka have to pay some performance/complexity overhead of dealing with the possibility of a framer even though that's built into the protocol?

Author: Hm, I see your point. The Kafka sink would still be able to add a `SerializerConfig` only, to build the `Serializer` without wrapping it in an `Encoder`.

Otherwise could be worth to do some micro benchmarks for how much of an impact going through one no-op framing operation has. Might not be significant, as the cost only occurs once per frame.

Author: Clarified in d7c9cbfe47a40dcc035fab4aea62037c3914647a.
